torch/csrc/autograd/record_function.cpp

```diff
@@ -137,6 +138,18 @@ bool hasNonSampledCallbacks() {
   return manager().hasNonSampledCallbacks();
 }
 
+void runBeforeCallbacks(RecordFunction* rf, const std::string& funcName) {
+  TORCH_INTERNAL_ASSERT(
+      rf != nullptr, "Passed in RecordFunction cannot be null.");
+  if (hasCallbacks()) {
+    auto run_samples = shouldRunSampledCallbacks();
+    if (run_samples || hasNonSampledCallbacks()) {
+      rf->setRunSampled(run_samples);
+      rf->before(funcName);
```

Comment:
do we need to worry about `sequence_nr` here?

Reply:
I don't think so, from the docs it looks like:
```
`sequence_nr` is
  /// a (currently THE) hint to prioritization in the backward() pass, with
  /// higher sequence numbers prioritized before lower sequence numbers.
```